fendermint/vm/topdown/src/sync/pointers.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::BlockHeight;
+use std::fmt::{Display, Formatter};
+
+/// There are three pointers, each refers to a block height, when syncing with parent. As Lotus has
+/// delayed execution and null round, we need to ensure the topdown messages and validator
+/// changes polled are indeed finalized and executed. The following three pointers are introduced:
+///     - tail: The latest block height in cache that is finalized and executed
+///     - to_confirm: The next block height in cache to be confirmed executed, could be None
+///     - head: The latest block height fetched in cache, finalized but may not be executed.
+///
+/// Say we have block chain as follows:
+/// NonNullBlock(1) -> NonNullBlock(2) -> NullBlock(3) -> NonNullBlock(4) -> NullBlock(5) -> NonNullBlock(6)
+/// and block height 1 is the previously finalized and executed block height.
+///
+/// At the beginning, tail == head == 1 and to_confirm == None. With a new block height fetched,
+/// `head = 2`. Since height at 2 is not a null block, `to_confirm = Some(2)`, because we cannot be sure
+/// block 2 has executed yet. When a new block is fetched, `head = 3`. Since head is a null block, we
+/// cannot confirm block height 2. When `head = 4`, it's not a null block, we can confirm block 2 is
+/// executed (also with some checks to ensure no reorg has occurred). We fetch block 2's data and set
+/// `tail = 2`, `to_confirm = Some(4)`. Then height 2 is ready to be proposed.
+#[derive(Clone, Debug)]
+pub(crate) struct SyncPointers {
```

Comment:
Thank you for the detailed example and description :pray: 

I don't understand why we need this `to_confirm` at all; when we have a non-null height we know that everything _before_ has been executed, and regardless of whether we know that some block has a follow-up that "confirms" the execution results, we should refrain from using it. That way everyone is on equal footing and there is no reason to wait any new blocks.

Reply:
when we have a non-null height we know that everything before has been executed, this is reflected in the `tail` variable. And `tail` is actually the latest height in cache. So when proposing, everything up till `tail` will be available for proposal. But the thing is, if there is a new block that is non-null, but how are we going to populate the cache? We don't know which previous block's data should we pull? If we do this, there will be empty data for non-null blocks. If we just wait, this won't happen.

If you are suggesting we can remove `to_confirm` and replace it with `tail`, I think that's a good simplification.

Reply:
To be honest I don't really want to think in the terms of tails and heads and waiting for confirms. 

What I have been proposing for a week is much simpler: as soon as you see a block, you know that its _parent_ has been executed, so when we finalize a given block, let's limit ourselves to the events of its parent. 

If we did not have null blocks, this would be simply the previous height. With null blocks we have to find the parent. But it's already in the cache, or it's already in the parent chain; the important thing is there is no situation where we have to wait for anything to happen.